packages/devtools/src/common/promise.ts

```diff
@@ -72,3 +73,131 @@ export const firstFactory =
     <TInput extends unknown[], TOutput>(...factories: Factory<TInput, TOutput>[]): Factory<TInput, TOutput> =>
     async (...input) =>
         await first(factories.map((factory) => () => factory(...input)))
+
+/**
+ * RetryStrategy is a way of having a fine grain control
+ * over execution of retried function.
+ *
+ * It will be executed on every failed attempt and has the ability to modify the
+ * input originally passed to the retried function.
+ *
+ * In its simplest form, it will either return `true` (to retry again) or `false` (stop retrying).
+ *
+ * In its advanced form, it can use the parameters passed to it to create
+ * a new set of arguments passed to the function being retried:
+ *
+ * ```
+ * // As a simple example let's consider a function
+ * // whose argument is the amount of money we want to pay for a service
+ * const functionThatCanFail = (money: number): Promise<void> => { ... }
+ *
+ * // We can create a strategy that will keep adding 1 to the amount of money
+ * const strategy: RetryStrategy<[money: number]> = (attempt, error, [previousMoney]) => [previousMoney + 1]
+ *
+ * // Or we can create a strategy that will adjust the money based on the initial value
+ * //
+ * // In this made up case it will take the original amount and will add 2 for every failed attempt
+ * const strategy: RetryStrategy<[money: number]> = (attempt, error, _, [originalMoney]) => [originalMoney + attempt * 2]
```

Comment:
because originalMoney is the name of the parameter each time, it seems this actually compounds right?
So its not like 10, 12, 14,... it would be 10, 12, 16, ...

Reply:
Fixed the code comments here not to confuse people - the `_` parameter placeholder has been replaced